experimental/attestation_common/src/certificate.rs

```diff
@@ -0,0 +1,317 @@
+// https://docs.rs/openssl/0.10.33/openssl/bn/struct.BigNum.html#method.rand
+const SERIAL_NUMBER_SIZE: i32 = 159;
+const CERTIFICATE_EXPIRATION_INTERVAL_IN_DAYS: u32 = 1;
+const DEFAULT_DNS_NAME: &str = "localhost";
+
+/// Indicator whether to add a custom TEE extension to a certificate.
+#[derive(PartialEq)]
+pub enum AddTeeExtension {
+    /// Parameter contains a PEM encoded X.509 certificate that signs TEE firmware key.
+    Yes(Vec<u8>),
+    No,
+}
+
+/// Convenience structure for generating X.509 certificates.
+/// https://tools.ietf.org/html/rfc5280
+pub struct CertificateAuthority {
+    pub key_pair: PKey<Private>,
+    pub root_certificate: X509,
+}
+
+impl CertificateAuthority {
+    /// Generates a root X.509 certificate and a corresponding private/public key pair.
+    /// `add_tee_extension` indicates whether to add a custom extension containing a TEE report to
+    /// the root certificate.
+    pub fn create(add_tee_extension: AddTeeExtension) -> anyhow::Result<Self> {
+        let key_pair = CertificateAuthority::generate_key_pair()?;
+        let root_certificate =
+            CertificateAuthority::generate_root_certificate(&key_pair, add_tee_extension)?;
+        Ok(Self {
+            key_pair,
+            root_certificate,
+        })
+    }
+
+    /// Generates RSA private/public key pair.
+    fn generate_key_pair() -> anyhow::Result<PKey<Private>> {
+        let rsa = Rsa::generate(RSA_KEY_SIZE).context("Couldn't generate RSA key")?;
+        PKey::from_rsa(rsa).context("Couldn't parse RSA key")
+    }
+
+    /// Creates a root X.509 certificate.
+    fn generate_root_certificate(
+        key_pair: &PKey<Private>,
+        add_tee_extension: AddTeeExtension,
+    ) -> anyhow::Result<X509> {
+        info!("Generating root certificate");
+
+        let mut builder = CertificateBuilder::create()?;
+        builder.set_version(CERTIFICATE_VERSION)?;
+        builder.set_serial_number(SERIAL_NUMBER_SIZE)?;
+        builder.set_name()?;
+        builder.set_public_key(key_pair)?;
+        builder.set_expiration_interval(CERTIFICATE_EXPIRATION_INTERVAL_IN_DAYS)?;
+
+        builder.add_basic_constraints_extension(true)?;
+        builder.add_key_usage_extension(true)?;
+        builder.add_subject_key_identifier_extension(None)?;
+        builder.add_subject_alt_name_extension()?;
+
+        // Bind the certificate to the TEE firmware using an X509 TEE extension.
+        if let AddTeeExtension::Yes(tee_certificate) = add_tee_extension {
+            builder.add_tee_extension(key_pair, tee_certificate)?;
+        }
+
+        let certificate = builder.build(key_pair)?;
+        Ok(certificate)
+    }
+
+    /// Generates an X509 certificate based on the certificate signing `request`.
+    /// `add_tee_extension` indicates whether to add a custom extension containing a TEE report.
+    pub fn sign_certificate(
+        &self,
+        request: X509Req,
+        add_tee_extension: AddTeeExtension,
```

Comment:
Is there a relation between this parameter and `add_tee_extension` in the `create` method above? I mean is any combination of the values of these two parameters valid? For example, is it correct to have a root-certificate that does not have a TEE extension but use it for signing certificates that have TEE extension? 
Or actually, how is the TEE extension in the root_certificate used? (sorry if you have explained it before!)

Reply:
`add_tee_extension` is going to be added for a machine that wants to be remotely attested.
For example, if we use self-signed certificates - we will add this parameter to the root certificate.
And if we want to use Proxy Attestation, we will sign a new certificate (for backend server) using our root one (that doesn't need the parameter since it's not going to be remotely attested). And the new certificate will be created using the `add_tee_extension` and be sent to the backend server.

Reply:
I see. Thanks. So they are sort of mutually exclusive.